### ISSUE TITLE
Transparently ask for permissions on operations

### ISSUE THREAD

**Issue by orowith2os** [flatpak/xdg-desktop-portal#1037]
Some portal APIs, including the Camera portal, first require the developer to ask the user for permission to the device first, and then they can access them.

On macOS, this is handled transparently; if the developer tries to access the device and the user denies it, the whole operation fails. 

With the portals, accessing the device will still fail, but it won't ask the user for permission. The developer has to request permission first, and *then* access the device if that fails.

Making these aforementioned APIs ask the user for permission without explicit developer interaction would be useful.

We already have the [Response signal](https://flatpak.github.io/xdg-desktop-portal/#gdbus-signal-org-freedesktop-portal-Request.Response), so it would be a relatively minor effort to change to using this design. Many other portals already use this for reference.

List of portals using this design:

- The [ScreenCast](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.ScreenCast) portal. This may need more discussion, as well as viewing how macOS handles it and if this redesign should be done for it in the first place. 
- The [RemoteDesktop](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.RemoteDesktop) portal. Same thing as the ScreenCast portal, needs more discussion.
- The [Print](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.Print) portal. Honestly, the design here is a bit confusing to me, as it appears there are two different methods, both doing *very* similar things. But worth bringing up here, as I see a bit of this design in the link between `PreparePrint()` and `Print()`.
- The [Location](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.Location) portal. 
- The [GlobalShortcuts](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.GlobalShortcuts) portal. See the ScreenCast portal.
- The [DynamicLauncher](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.DynamicLauncher) portal. The process of setting up a desktop file could be simplified into one method.

- The [Camera](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.Camera) portal. Could be simplified to just AccessCamera, but:
- The [Device](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.Device) portal. This currently only has AccessDevice, and the ability to actually *access* those devices is [being worked on](https://github.com/flatpak/xdg-desktop-portal/pull/659). As this portal is a bit more generic, it may need more discussion, and some better API decisions regarding accessing several devices. This could be expanded to be more of an actual portal, and not something meant to be used by services, thus succeeding the Camera portal.

Set general guidelines inside of a document for making new portals would be useful to make portals more like this suggestion, as well as show what services are suitable for portals and what aren't, as portals now also behave as a general desktop-integration API too.

**Comment by GeorgesStavracas** [comment]
After the discussion on the Matrix channel a while ago, the rough consensus is that we don't want to change how permissions work now, specifically the two-step process is sufficient for now and there's no big reason to change it at this point. So this is out of scope.

**Comment by liangqi** [comment]
See also Permission API implementation for Linux, https://bugreports.qt.io/browse/QTBUG-122555

**Comment by TingPing** [comment]
@liangqi I'm not familiar with Qt, but at a glance their APIs match the flow of portals.

`checkPermission()`, `requestPermission()`, and then usage are separate steps?

**Comment by liangqi** [comment]
> `checkPermission()`, `requestPermission()`, and then usage are separate steps?

Do you mean portals provide those functions? Thanks.

**Comment by liangqi** [comment]
Perhaps PermissionStore could be used? https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.impl.portal.PermissionStore.html#

**Comment by GeorgesStavracas** [comment]
Apps should now use any backend portal interfaces (i.e. anything with the `org.freedesktop.impl.portal.` prefix)

**Comment by TingPing** [comment]
> Do you mean portals provide those functions? Thanks.

Basically every portal you request permissions and then you can use it.

Some of them you can know ahead of time if you have permissions because you hold a token that was used to grant it last time.

**Comment by liangqi** [comment]
> Perhaps PermissionStore could be used? https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.impl.portal.PermissionStore.html#

That's linked from official xdg portal doc, https://flatpak.github.io/xdg-desktop-portal/ -> https://flatpak.github.io/xdg-desktop-portal/docs/index.html

**Comment by TingPing** [comment]
Yes its an interface for backends, the section has a whole header about it. Not applications.
